active-rfcs/0000-sfc-script-setup.md

```diff
@@ -239,6 +258,77 @@ export function setup() {
 export default { setup }
 ```
 
+## Transform API
+
+The `@vue/compiler-sfc` package exposes the `compileScript` method for processing `<script setup>`:
+
+```js
+import { parse, compileScript } from '@vue/compiler-sfc'
+
+const descriptor = parse(`...`)
+
+if (descriptor.scriptSetup) {
+  const result = compileScript(descriptor) // returns SFCScriptBlock
+  console.log(result.code)
+  console.log(result.bindings) // see next section
+}
+```
+
+The compilation requires the entire descriptor to be provided, and the resulting code will include sources from both `<script setup>` and normal `<script>` (if present). It is the higher level tools' (e.g. `vite` or `vue-loader`) responsibility to properly assemble the compiled output.
+
+## Template Binding Optimization
+
+The `SFCScriptBlock` returned by `compiledScript` also exposes a `bindings` object, which is the exported binding metadata gathered during the compilation. For example, given the following `<script setup>`:
+
+```vue
+<script setup="props">
+export const foo = 1
+
+export default {
+  props: ['bar']
+}
+</script>
+```
+
+The `bindings` object will be:
```

Comment:
@yyx990803 I looked up the `BindingMetadata` typedef in compiler-core. Following types are defined:  `'data' | 'props' | 'setup' | 'options'`

I think an additional value `A: 'import'`, for `export { A } from './a'` could lead to interesting optimizations.

The compiler would then be free to reference the import directly, without reactivity and bypassing the setup state for references to `A`.

That's particularly interesting for local components (or directives).
This RFC also looks for a syntax to import local components. With the ability to recognize imports (constants) and reference them directly, it doesn't need one!

If the template does `<MyButton />` and the script setup contains an `export { MyButton } from './my-button'` identified as such, the compiler could produce the equivalent of `h(MyButton, ...)`.

Without this knowledge, the local component would suffer two drawbacks:
- going through the reactivity layer to access `MyButton` on the setup object.
- assuming that `MyButton` is a variable that could change, so the component is dynamic and precludes the static optimizations.

Reply:
1. The thing is that the template and script parts of an SFC are typically executed as separate modules to
    1. allow each having their own loader pipelines (webpack specific)
    2. allow template to be individually hot-reloaded (thus preserving component state).

Regarding the drawbacks:

- Component access only goes through the setup object, which is not a full reactive object (it's a proxy that only checks for ref unwrapping, so the cost is fairly cheap).

- `<MyButton/>` directly compiles to `h($setup.MyButton)`, so there is no dynamic assumptions here.

Reply:
> `<MyButton/>` directly compiles to h($setup.MyButton), so there is no dynamic assumptions here.

I think I'm missing a piece here.

Setup properties can mutate, right? Do you know/assume `$setup.MyButton` is a constant rather than a reactive value that will change?

If you assume `MyButton` can mutate, then isn't the code equivalent to `<component :is="MyButton">`, which precludes some optimizations compared to a static `<MyButton>`?
https://vue-next-template-explorer.netlify.app/#%7B%22src%22%3A%22%3Ccomponent%20%3Ais%3D%5C%22xy%5C%22%20%2F%3E%5Cr%5Cn%3Cmy-xy%20%2F%3E%22%2C%22options%22%3A%7B%22mode%22%3A%22module%22%2C%22prefixIdentifiers%22%3Afalse%2C%22optimizeImports%22%3Afalse%2C%22hoistStatic%22%3Afalse%2C%22cacheHandlers%22%3Afalse%2C%22scopeId%22%3Anull%2C%22ssrCssVars%22%3A%22%7B%20color%20%7D%22%2C%22bindingMetadata%22%3A%7B%22TestComponent%22%3A%22setup%22%2C%22foo%22%3A%22setup%22%2C%22bar%22%3A%22props%22%7D%2C%22optimizeBindings%22%3Afalse%7D%7D

Or does the syntax `<MyButton>` imply that it must be static, even though it comes from `setup`? 
In this case, what happens if that assumption is violated by user, do you emit a warning in DEV?